gr-digital/include/gnuradio/digital/decision_feedback_equalizer.h

```diff
@@ -0,0 +1,85 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2020 Free Software Foundation, Inc.
+ *
+ * This file is part of GNU Radio
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ *
+ */
+
+#ifndef INCLUDED_DIGITAL_DECISION_FEEDBACK_EQUALIZER_H
+#define INCLUDED_DIGITAL_DECISION_FEEDBACK_EQUALIZER_H
+
+#include <gnuradio/digital/adaptive_algorithm.h>
+#include <gnuradio/digital/api.h>
+#include <gnuradio/sync_decimator.h>
+
+namespace gr {
+namespace digital {
+
+/*!
+ * \brief Decision Feedback Equalizer
+ * \ingroup equalizers
+ *
+ */
+class DIGITAL_API decision_feedback_equalizer : virtual public gr::sync_decimator
+{
+public:
+    typedef boost::shared_ptr<decision_feedback_equalizer> sptr;
+
+    /*!
+     * \brief Return a shared_ptr to a new instance of
+     * gr::digital::decision_feedback_equalizer.
+     *
+     * The Decision Feedback Equalizer block equalizes the incoming signal using an
+     * FIR filter representing both the forward and feedback taps
+     *
+     * If provided with a training sequence and a training start tag, data aided
+     * equalization will be performed starting with the tagged sample. If training-based
+     * equalization is active and the training sequence ends, then optionally decision
+     * directed equalization will be performed given the adapt_after_training If no
+     * training sequence or no tag is provided, decision directed equalization will be
+     * performed
+     * This equalizer decimates to the symbol rate according to the samples per symbol
+     * param
+     *
+     * \param num_taps_forward  Number of feedforward taps for the FIR filter
+     * \param num_taps_feedback  Number of feedback taps for the FIR filter
+     * \param sps  int - Samples per Symbol
+     * \param alg  Adaptive algorithm object
+     * \param training_sequence  Sequence of samples that will be used to train the
+     * equalizer.  Provide empty vector to default to DD equalizer
+     * \param adapt_after_training  bool - set true to continue DD training after training
+     * sequence has been used up
+     * \param training_start_tag  string - tag that indicates the start
+     * of the training sequence in the incoming data
+     */
+    static sptr
+    make(int num_taps_forward,
+         int num_taps_feedback,
+         int sps,
+         adaptive_algorithm_sptr alg,
+         bool adapt_after_training = true,
+         std::vector<gr_complex> training_sequence = std::vector<gr_complex>(),
+         const std::string& training_start_tag = "");
+
+    virtual void set_taps(const std::vector<gr_complex>& taps) = 0;
+    virtual std::vector<gr_complex> taps() const = 0;
+
+    // Public "work" function - allow signal processing to be called externally
```

Comment:
Fair question.  There a couple of reasons for structuring the block this way:
1. The signal processing can be called outside a gnu radio flowgraph.  i could simply instantiate an equalizer object in python or c++ and call the equalize method directly.  state would be maintained, but doesn't necessarily have to be performed in a streaming context.  Some blocks have done this with the "kernel" namespaces.  
2. Makes it easier to also add pdu or tagged stream versions wrapping the same signal processing

Reply:
OK. That's a good reason. Though, I'd recommend to go with the kernel structure. This makes it actually usable outside GR. Otherwise, you'd need all the block infrastructure if just want to use the kernel. That would be a clean way to separate functionality from the interface.

Reply:
I see the benefit of the kernel namespace outside of GR - for this implementation, the equalize() function still relies on tags which are GR specific so not sure if the benefits of the kernel namespace would be fully realized.  Having the equalize() function broken off does bring some benefits - I'd like to leave it as is if that is ok.